theano/tensor/var.py

```diff
@@ -154,6 +154,14 @@ def __sub__(self, other):
         except (NotImplementedError, AsTensorError):
             return NotImplemented
 
+    def __matmul__(self, other):
+        # See explanation in __add__ for the error catched
+        # and the return value in that case
+        try:
+            return theano.tensor.dot(self, other)
```

Comment:
Hmmm...  It seems you are right, tensor does indeed have different behaviour than intended in PEP 465.
the @ operator covers the matmult operator in numpy, an operator that for all i know does not exist in theano atm.

from http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.matmul.html
```
Matrix product of two arrays.

The behavior depends on the arguments in the following way.

If both arguments are 2-D they are multiplied like conventional matrices.
If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.
If the first argument is 1-D, it is promoted to a matrix by prepending a 1 to its dimensions. After matrix multiplication the prepended 1 is removed.
If the second argument is 1-D, it is promoted to a matrix by appending a 1 to its dimensions. After matrix multiplication the appended 1 is removed.
Multiplication by a scalar is not allowed, use * instead. Note that multiplying a stack of matrices with a vector will result in a stack of vectors, but matmul will not recognize it as such.

matmul differs from dot in two important ways.

Multiplication by scalars is not allowed.
Stacks of matrices are broadcast together as if the matrices were elements.
```

There is however implemented functionality for N-D, N <= 3 as `basic.batched_dot` will work for N == 3
I will look into an implementation of tensor.basic.matmult that mirrors the numpy one